drivers/net/ethernet/renesas/rswitch.c

```diff
@@ -2435,16 +2452,35 @@ static int rswitch_start_xmit(struct sk_buff *skb, struct net_device *ndev)
 	unsigned long flags;
 	struct rswitch_gwca_chain *c = rdev->tx_chain;
 	int i, num_desc, pkt_len, size;
+	int rx_chain_free = RX_RING_SIZE;
+	u64 txed;
 
 	spin_lock_irqsave(&rdev->lock, flags);
 
 	num_desc = skb->len % MAX_DESC_SZ ? skb->len / MAX_DESC_SZ + 1 : skb->len / MAX_DESC_SZ;
+	if (rdev->is_vmq) {
+		if (rswitch_is_front_dev(rdev)) {
+			rx_chain_free = RX_RING_SIZE -
+					(READ_ONCE(rdev->vmq_info->front_tx) -
+					READ_ONCE(rdev->vmq_info->back_rx));
+		} else {
+			rx_chain_free = RX_RING_SIZE -
+				        (READ_ONCE(rdev->vmq_info->back_tx) -
+					READ_ONCE(rdev->vmq_info->front_rx));
+		}
+	}
 
 	if (c->cur - c->dirty > c->num_ring - num_desc) {
 		netif_stop_subqueue(ndev, 0);
 		ret = NETDEV_TX_BUSY;
 		goto out;
 	}
+	
+	if (num_desc > rx_chain_free) {
+		curr_time = ktime_get();
+		ret = NETDEV_TX_BUSY;
```

Comment:
Seems it is correct to call `netif_stop_queue/netif_stop_subqueue` here to stop upper layers of network stack to bother us. 

Question is - when we should call `netif_start_subqueue`? It would be really nice to call it when, say, 1/2 or 3/4 of TX chain is free. IIRC, GWCA sends TX interrupt when chain is completely empty, which is late a bit. I wonder - is it possible to flag TX descriptor with some attribute to force GWCA trigger an interrupt when it processed it? In this way we can flag every 64th (or 128th or something like that) TX descriptor and get steady stream of TX interrupts...

On other hand, we really interested in interrupt from other side (e.g. from backend). Can you try the following: send TX done event to DomU every time DomD exits its `rx_poll` function?